judge-d-agent/src/test/functional/com/hltech/judged/agent/K8sServiceLocatorFT.groovy

```diff
@@ -0,0 +1,122 @@
+import static org.awaitility.Awaitility.await
+
+@SpringBootTest
+@ActiveProfiles(["test", "kubernetes"])
+class K8sServiceLocatorFT extends Specification {
+
+    @Shared
+    KubernetesMockServer k8sServer
+
+    static WireMockRule wireMockRule
+
+    @Autowired
+    KubernetesClient kubernetesClient
+
+    def setupSpec() {
+        k8sServer = new KubernetesMockServer()
+        k8sServer.init()
+        System.setProperty(Config.KUBERNETES_MASTER_SYSTEM_PROPERTY, "https://${k8sServer.hostName}:$k8sServer.port/")
+        System.setProperty(Config.KUBERNETES_TRUST_CERT_SYSTEM_PROPERTY, "true")
+        wireMockRule = new WireMockRule(8080)
+        wireMockRule.start()
+    }
+
+    def "should send update services message with expected names and versions"() {
+        given: "expected namespaces"
+            k8sServer.expect().get().withPath("/api/v1/namespaces")
+                .andReturn(200, new NamespaceListBuilder()
+                    .addNewItem()
+                    .withNewMetadata()
+                    .withName("firstnamespace")
+                    .endMetadata()
+                    .and()
+                    .addNewItem()
+                    .withNewMetadata()
+                    .withName("secondnamespace")
+                    .endMetadata()
+                    .and()
+                    .build())
+                .always()
+
+        and: "expected pods"
+            PodList expectedPodList = new PodListBuilder()
+                .withItems(
+                    buildPod("firstnamespace", 1, "78d588f9cc"),
+                    buildPod("firstnamespace", 2, "54d576f9dh"),
+                    buildPod("secondnamespace", 3, "98d500g9df")
+                ).build()
+
+            k8sServer.expect().get().withPath("/api/v1/pods?labelSelector=app")
+                .andReturn(HttpURLConnection.HTTP_OK, expectedPodList)
+                .always()
+
+        and: "expected message to publish"
+            def expectedRequestBody = """
+                [
+                    {
+                        "name": "test_app_1",
+                        "version": "78d588f9cc"
+                    },
+                    {
+                        "name": "test_app_2",
+                        "version": "54d576f9dh"
+                    },
+                    {
+                        "name": "test_app_3",
+                        "version": "98d500g9df"
+                    }
+                ]
+            """
+
+            wireMockRule.stubFor(
+                put(urlPathMatching("/environments/test"))
+                    .withRequestBody(equalToJson(expectedRequestBody, true, false))
+                    .willReturn(ok())
+            )
+
+
+        expect: "update services message is sent with expected services names and versions"
+            await().atMost(15, TimeUnit.SECONDS).until({ wireMockRule.getServeEvents().requests.size() == 1 })
```

Comment:
please bind waiting time to the period set for judge-d agent updates so that when the period is changed, we will not make it give random results

Reply:
done